security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,236 @@
+				caSecret.Data[caPrivateKeyID], nil, caSecret.Data[caCertID]); err != nil {
+				rootCertRotatorLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
+			} else {
+				rootCertRotatorLog.Info("Successfully reloaded root cert into KeyCertBundle.")
+			}
+			certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+			// Keep root certificate in configmap in sync with the root certificate in istio-ca-secret.
+			if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+				certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+				rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+					"to configmap (%s). Citadel agents will not be able to connect.",
+					err.Error())
+			} else {
+				rootCertRotatorLog.Info("Root certificate is updated into configmap.")
+			}
+		}
+		return
+	}
+
+	rootCertRotatorLog.Infof("Refresh root certificate, root cert is about to expire: %s", err.Error())
+	options := util.CertOptions{
+		TTL:           rotator.config.caCertTTL,
+		SignerPrivPem: caSecret.Data[caPrivateKeyID],
+		Org:           rotator.config.org,
+		IsCA:          true,
+		IsSelfSigned:  true,
+		RSAKeySize:    caKeySize,
+		IsDualUse:     rotator.config.dualUse,
+	}
+	pemCert, pemKey, ckErr := util.GenRootCertFromExistingKey(options)
+	if ckErr != nil {
+		rootCertRotatorLog.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+		return
+	}
+
+	oldCaCert := caSecret.Data[caCertID]
+	oldCaPrivateKey := caSecret.Data[caPrivateKeyID]
+	if rollback, err := rotator.updateRootCertificate(caSecret, true, pemCert, pemKey); err != nil {
+		if !rollback {
+			rootCertRotatorLog.Errorf("Failed to roll forward root certificate (error: %s). "+
+				"Abort new root certificate", err.Error())
+			return
+		}
+		// caSecret is out-of-date. Need to load the latest istio-ca-secret to roll back root certificate.
+		_, err = rotator.updateRootCertificate(nil, false, oldCaCert, oldCaPrivateKey)
+		if err != nil {
+			rootCertRotatorLog.Errorf("Failed to roll backward root certificate (error: %s).", err.Error())
+		}
+		return
+	}
+	rootCertRotatorLog.Info("Root certificate rotation is completed successfully.")
+}
+
+// updateRootCertificate updates root certificate in istio-ca-secret, keycertbundle and configmap. It takes a scrt
+// object, cert, and key, and a flag rollForward indicating whether this update is to roll forward root certificate or
+// to roll backward.
+// updateRootCertificate returns error when any step is failed, and a flag indicating whether a rollback is required.
+// Only when rollForward is true and failure happens, the returned rollback flag is true.
+func (rotator *SelfSignedCARootCertRotator) updateRootCertificate(caSecret *v1.Secret, rollForward bool, cert, key []byte) (bool, error) {
+	var err error
+	if caSecret == nil {
```

Comment:
I would just require `caSecret` to be not `nil` here, and I would make a copy of the given `caSecret` before modifying the `caCertID` and `caPrivateKeyID` keys below.
That way, in case of retry you can pass the original `caSecret` (not `nil`) and avoid the extra cost of calling `LoadCASecretWithRetry`.

Reply:
The LoadCASecretWithRetry() is to get the latest caSecret for rollback. 
Once we pass new root cert into caSecret and update istio-ca-secret with caSecret succesfully, that caSecret object is outdated. If we decide to rollback the root cert, we need to load istio-ca-secret into a new caSecret object, pass old root cert into that caSecret object, and update istio-ca-secret. Using an outdated caSecret object to update istio-ca-secret will fail, as the k8s API server rejects this object. I have done experiments to verify this.